clean_air/util/dataframe_converter.py

```diff
@@ -0,0 +1,148 @@
+        data_pd = []
+        geom = []
+        i = 0
+        for x in x_coord.points:
+            # Reset the y-coord index to 0 each time we loop over the x-coord:
+            j = 0
+            for y in y_coord.points:
+                x_coords_pd.append(x)
+                y_coords_pd.append(y)
+                data_pd.append(sub_cube.data[i, j])
+                geom.append(Point(x, y))
+                # i and j represent the index values of the coords, so
+                # they need to stop counting at the end of the coord arrays.
+                if j < len(y_coord.points) - 1:
+                    j += 1
+            if i < len(x_coord.points) - 1:
+                i += 1
+
+        # Now construct pandas dataframe object for this sub_cube:
+        pandas_df = pd.DataFrame({'x_coord': x_coords_pd,
+                                  'y_coord': y_coords_pd,
+                                  'data': data_pd})
+
+        # Then make geometry in correct form (which is a list of
+        # shapely.geometry.Points objects) and convert to GeoDataFrame:
+        geo_df = GeoDataFrame(pandas_df, geometry=geom)
+        dataframes.append(geo_df)
+
+    return dataframes
+
+
+def convert_to_geodf(cube, restitch=True):
+    """
+    Callable for converting iris-style cubes into geopandas dataframes.
+
+    Note: This conversion is only possible for 2-dimensional cubes, so input
+    cubes with more than two dimensions will be sliced up before conversion
+    and then restitched if necessary.  Cubes with less than two dimensions
+    will be converted to a series instead.
+
+    :param cube: input cube.
+    :param restitch: Boolean representing whether to concatenate the
+    dataframes after conversion to pandas arrays.
+    :return: pandas-style dataframe(s) or series, depending on input (see note).
+    """
+    geodataframes = series = geodataframe = None
+    if len(cube.dim_coords) == 2:
+        restitch = False
+
+    # For cubes with more than two dim coords:
+    if len(cube.dim_coords) >= 2:
+        x_coord, y_coord = get_xy_coords(cube)
+        # If there are more than two coords, we need to look for standard
+        # xy dimension coords and use them to slice the cube up:
+        geodataframes = _make_geo(cube, x_coord, y_coord)
+
+        # Now we have a set of 2D geopandas dataframes which must be
+        # stitched back together if required
+        if restitch:
+            geodataframe = pd.concat(geodataframes)
+
+    # For cubes with less than 2 dim coords (warn then convert to series):
+    elif len(cube.dim_coords) < 2:
+        warnings.warn("Converting to non-geographical series instead of "
+                      "geodataframe as there is only one coordinate.")
+        series = as_series(cube)
+
+    if len(cube.dim_coords) >= 2:
+        if restitch is True:
+            return geodataframe
+        else:
+            return geodataframes
+    elif len(cube.dim_coords) < 2:
+        return series
```

Comment:
The reason I didn't do this is because I thought it was bad practise to put a return statement in the middle of a function, but I'm not sure about that and I do agree with you, so I'll rearrange this and we can have another look at it.

Reply:
Yeah seems fine, no complaints from pycharm and it looks and feels much better.